### PR TITLE
Bug 1755115: do not schedule imports for streams with sha/digest references whose …

### PR DESCRIPTION
**Pull request by gabemontero**
…import has completed

per instructions from @smarterclayton  ... the api needs to be  “level driven”

with this change, we can use scheduled import to enact retry for imagestreams with dockerimage refs that container the canonical sha/digest, but once they are successful, no longer try to import, as the sha/digest is the final outcome

/assign @bparees 
/assign @adambkaplan 
/assign @dmage

@openshift/openshift-team-developer-experience FYI

### CODE REVIEW

**Comment by gabemontero** [comment]
this change does not require it, but users will ultimately need https://github.com/openshift/openshift-apiserver/pull/28 merged to be able to leverage this new path

File: pkg/image/controller/scheduled_image_controller.go [hunk 1]
```diff
@@ -189,6 +190,13 @@ func resetScheduledTags(stream *imagev1.ImageStream) {
 func needsScheduling(stream *imagev1.ImageStream) bool {
 	for _, tagRef := range stream.Spec.Tags {
 		if tagImportable(tagRef) && tagRef.ImportPolicy.Scheduled {
+			if ref, err := imageref.Parse(tagRef.From.Name); err != nil && len(tagRef.From.Name) > 0 {
+				return false
```
Comment (bparees) [inline pkg/image/controller/scheduled_image_controller.go]: shouldn't the validation code prevent this from happening?  (it's the same check the validation is doing, right?)

Comment (bparees) [reply pkg/image/controller/scheduled_image_controller.go]: i guess if you have to parse it anyway to get the id it makes sense to respect the error, if any though.

File: pkg/image/controller/scheduled_image_controller_test.go [hunk 2]
```diff
@@ -124,3 +124,124 @@ func TestScheduledImport(t *testing.T) {
...
+					Name: "test", Namespace: "other", UID: "1", ResourceVersion: "1",
+					Annotations: map[string]string{imagev1.DockerImageRepositoryCheckAnnotation: "done"},
+					Generation:  1,
+				},
+				Spec: imagev1.ImageStreamSpec{
+					Tags: []imagev1.TagReference{
+						{
+							Name:         "default",
+							From:         &corev1.ObjectReference{Kind: "DockerImage", Name: "abc@sha256:3c87c572822935df60f0f5d3665bd376841a7fcfeb806b5f212de6a00e9a7b25"},
+							Generation:   &one,
```
Comment (bparees) [inline pkg/image/controller/scheduled_image_controller_test.go]: this test would be better if the spec generation was newer than the status generation, to confirm we still don't import a SHA tag that is already successfully imported, right?

**Comment by bparees** [comment]
i think there are some issues here....

needsScheduling looks at the imagestream as a whole to make a decision, so if any tag within it needs to be scheduled, the whole imagestream will be scheduled.

Then when the schedule actually comes up, resetScheduledTags is called to setup the specific tags for re-import.  So resetScheduledTags needs to also avoid resetting SHA tags that are already imported.  (probably needs to make the same needsImport check)

File: pkg/image/controller/scheduled_image_controller_test.go (hunk 2 above)
Comment (bparees) [reply pkg/image/controller/scheduled_image_controller_test.go]: ok on further study of the code, these values matching is our signal that the import is done/successful.  But i'm still wondering how the spec.generation value gets updated...i'm wondering if someone updates the imagestream, whether spec.generation gets updated and that triggers another round of imports.  If we can confirm that's not the case, then i think it's reasonable to assume that "once imported successfully, spec.generation will never exceed spec.status, for a sha based tag, as long as our schedule import avoidance logic is sound"

File: pkg/image/controller/scheduled_image_controller.go (hunk 1 above)
Comment (gabemontero) [reply pkg/image/controller/scheduled_image_controller.go]: that is where I landed on this one as well

File: pkg/image/controller/scheduled_image_controller_test.go (hunk 2 above)
Comment (gabemontero) [reply pkg/image/controller/scheduled_image_controller_test.go]: Ok it looks like the logic which says a tagRef has changed are here:  https://github.com/openshift/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestream/strategy.go#L375-L423

That logic above then dictates whether the tagRef.generation is updated at https://github.com/openshift/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestream/strategy.go#L217-L228

I believe all the conditions at https://github.com/openshift/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestream/strategy.go#L375-L423 would be valid reasons for re-importing in our case, but let me know if you disagree @bparees

**Comment by gabemontero** [comment]
@bparees I've pushed an update to address the problem you found in https://github.com/openshift/openshift-controller-manager/pull/29#issuecomment-535502573

File: pkg/image/controller/scheduled_image_controller_test.go (hunk 2 above)
Comment (bparees) [reply pkg/image/controller/scheduled_image_controller_test.go]: yup, i agree, thanks for digging into it @gabemontero

File: pkg/image/controller/scheduled_image_controller.go [hunk 3]
```diff
@@ -180,6 +180,13 @@ func resetScheduledTags(stream *imagev1.ImageStream) {
 	next := stream.Generation + 1
 	for tag, tagRef := range stream.Spec.Tags {
 		if tagImportable(tagRef) && tagRef.ImportPolicy.Scheduled {
+			if ref, err := imageref.Parse(tagRef.From.Name); err != nil && len(tagRef.From.Name) > 0 {
+				continue
+			} else if len(ref.ID) > 0 && !tagNeedsImport(stream, tagRef, true) {
+				// ref.ID is set if this is a canonical, sha/digest ref;
+				// we allow scheduled import of those, but only until the initially succeed
+				continue
+			}
```
Comment (bparees) [inline pkg/image/controller/scheduled_image_controller.go]: i think it is still reasonable to also do this check in needsScheduling.  (if no tags in the imagestream end up needing scheduling, we save ourselves some work)

File: pkg/image/controller/scheduled_image_controller_test.go (hunk 2 above)
Comment (gabemontero) [reply pkg/image/controller/scheduled_image_controller_test.go]: cool thanks

File: pkg/image/controller/scheduled_image_controller.go (hunk 3 above)
Comment (gabemontero) [reply pkg/image/controller/scheduled_image_controller.go]: ok cool yeah that is why I left it there

Comment (bparees) [reply pkg/image/controller/scheduled_image_controller.go]: oh, i must have only been looking at the new commit, so i thought you moved it to here only.  nm.

**Comment by bparees** [comment]
squash + lgtm
/approve

**Comment by gabemontero** [comment]
squashed @bparees

**Comment by bparees** [comment]
/lgtm

let's make sure this(and the related changes) gets some good soak time before backporting to 4.2 though

File: pkg/image/controller/scheduled_image_controller.go [hunk 4]
```diff
@@ -189,6 +197,13 @@ func resetScheduledTags(stream *imagev1.ImageStream) {
 func needsScheduling(stream *imagev1.ImageStream) bool {
 	for _, tagRef := range stream.Spec.Tags {
 		if tagImportable(tagRef) && tagRef.ImportPolicy.Scheduled {
+			if ref, err := imageref.Parse(tagRef.From.Name); err != nil && len(tagRef.From.Name) > 0 {
+				return false
+			} else if len(ref.ID) > 0 && !tagNeedsImport(stream, tagRef, true) {
+				// ref.ID is set if this is a canonical, sha/digest ref;
+				// we allow scheduled import of those, but only until the initially succeed
+				return false
```
Comment (dmage) [inline pkg/image/controller/scheduled_image_controller.go]: What if we have two tags and the second one needs to be scheduled?

Comment (gabemontero) [reply pkg/image/controller/scheduled_image_controller.go]: good catch @dmage 

I'll queue up a follow up PR momentarily.

Comment (bparees) [reply pkg/image/controller/scheduled_image_controller.go]: yes, thanks @dmage, this should be a continue like is in resetScheduledTags.  (same for the err check above)

**Pull request merged**